walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
+    /**
+     * The timeTable collection holds TimeRanges which represent empty time slots for
+     * all the people invited
+     */
+    List<TimeRange> timeTable = new ArrayList<>();
+    List<TimeRange> optionalTimeTable = new ArrayList<>();
+    timeTable.add(TimeRange.WHOLE_DAY);
+    optionalTimeTable.add(TimeRange.WHOLE_DAY);
+
+    for (Event event : events) {
+      if (relevantEvent(event, mandatoryAttendees)) {
+        splitTimeTable(timeTable, event.getWhen());
+        splitTimeTable(optionalTimeTable, event.getWhen());
+      }
+
+      if (relevantEvent(event, optionalAttendees)) {
+        splitTimeTable(optionalTimeTable, event.getWhen());
+      }
+    }
+
+    return new TimeTablePair(timeTable, optionalTimeTable);
+  }
+
+  private Boolean relevantEvent(final Event event, final Collection<String> relevantPeople) {
+    for (String person : relevantPeople) {
+      if (event.getAttendees().contains(person)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private void splitTimeTable(final List<TimeRange> originalTimeTable,
+                              final TimeRange timeSplitter) {
+    /**
+     * Possible cases:
+     * 1. originalTimeTable:  [--------------------------]
+     *    splitter:                     [-----]
+     *    result:             [---------]     [----------]
+     *
+     * 2. originalTimeTable:  [------]   ...   [---------]
+     *    splitter:                [-------------]
+     *    result:             [----]             [-------]
+     *
+     * 3. originalTimeTable:  [------]   ...  [----------]
+     *    splitter:                     [-----]
+     *    result:             [------]        [----------]
+     *
+     * 4. originalTimeTable:  [------]           [-------]
+     *    splitter:                     [-----]
+     *    result:             [------]           [-------]
+     *
+     * Mention: for this method, it will only matter if the splitter is contained or not
+     */
+    if (originalTimeTable.isEmpty()) {
+      return;
+    }
+    int firstRelevantTimeslotIdx = TimeRange.lowerBound(originalTimeTable, timeSplitter);
+
+    if (originalTimeTable.get(firstRelevantTimeslotIdx).contains(timeSplitter)) {  // Case 1
+      splitAtTimeslot(originalTimeTable, timeSplitter, firstRelevantTimeslotIdx);
+    } else {  // Case 2, 3, 4
+      modifyOverlappingSlots(originalTimeTable, timeSplitter, firstRelevantTimeslotIdx);
+    }
+  }
+
+  private void splitAtTimeslot(final List<TimeRange> timeTable, final TimeRange splitter,
+                                                                final int splitIdx) {
+    TimeRange originalSlot = timeTable.remove(splitIdx);
+
+    if (originalSlot.end() > splitter.end()) {
+      timeTable.add(splitIdx,
+                    TimeRange.fromStartEnd(splitter.end(), originalSlot.end(), false));
+    }
+    
+    if (originalSlot.start() < splitter.start()) {
+      timeTable.add(splitIdx,
```

Comment:
Short comments sprinkled in the code, for example how inserting the start interval at the splitIdx here will insert it before the potential end interval, will assist reading the code faster.